File: lib/Refreshers/4chan/Refresher.ts
import consola from 'consola';
import type {
  Opaque,
  ValueOf,
} from 'type-fest';
import Boards from '../../Fetchers/4chan/Boards';
import Threads from '../../Fetchers/4chan/Threads';
import {
  callStorage,
} from '../../Redis/redis';
import type {
  Board,
  Post,
} from '../../Types/4chan/local';

const REDIS_KEY = 'refreshers:4chan:queue' as const;
const REDIS_ENTITY_SEPARATOR = '$' as const;

enum HandlerId {
  Boards = 'Boards',
  Threads = 'Threads',
}

type EntityId = Opaque<string, 'Entity ID'>;

const handlers = {
  [ HandlerId.Boards ]:
    async (): Promise<Board[]> =>
      await Boards.refreshCached(),
  [ HandlerId.Threads ]:
    async (board: EntityId): Promise<Post[]> =>
      await Threads.refreshCached(board),
};

class Entity {
  public readonly handlerId: HandlerId;

  public readonly entityId: EntityId;

  constructor(handlerId: Entity['handlerId'], entityId: string) {
    this.handlerId = handlerId;
    this.entityId = entityId as EntityId;
  }

  public encode(): string {
    return `${ this.handlerId }${ REDIS_ENTITY_SEPARATOR }${ this.entityId }`;
  }

  public static decode(redisEntity: string): Entity {
    const [ handlerId, ...entityIds ] = String(redisEntity).split(REDIS_ENTITY_SEPARATOR);
    const entityId = entityIds.join('$');

    return new this(
      handlerId as Entity['handlerId'],
      entityId as Entity['entityId'],
    );
  }
}

type ProcessResponse = {
  success: boolean;
  entity: Entity;
};

export class EntityStore {
  public static async pop(): Promise<Entity> {
    const redisEntity =
      await callStorage(
        'lpop',
        REDIS_KEY,
      )
    ;

    return Entity.decode(redisEntity);
  }

  public static async push(entity: Entity): Promise<void> {
    const redisEntity = entity.encode();

    await callStorage(
      'rpush',
      REDIS_KEY,
      redisEntity,
    );
  }

  public static async handle(redisEntity: Entity): Promise<ReturnType<ValueOf<typeof handlers>> | null> {
    const {
      handlerId,
      entityId,
    } = redisEntity;
    const handler = handlers[ handlerId ];

    if (!handler) {
      consola.error('Unknown handler', handlerId);
      return null;
    }

    return await handler(entityId);
  }

  public static async populate(): Promise<void> {
    const boards = await handlers[ HandlerId.Boards ]();

    await this.push(new Entity(HandlerId.Boards, 'all'));

    for (const { board } of boards) {
      await this.push(new Entity(HandlerId.Threads, board));
    }
  }

  public static async clean(): Promise<void> {
    await callStorage(
      'del',
      REDIS_KEY,
    );
  }

  public static async processOne(): Promise<ProcessResponse> {
    const entity = await this.pop();
    const res = await this.handle(entity);
    await this.push(entity);

    if (null === res) {
      consola.error('Something went wrong while processing entity', entity);
      return {
        success: false,
        entity,
      };
    }

    return {
      success: true,
      entity,
    };
  }

  public static async processAll(): Promise<ProcessResponse[]> {
    const {
      success,
      entity: firstEntity,
    } = await this.processOne();

    const entities = [
      {
        success,
        entity: firstEntity,
      },
    ];

    let currentEntity = new Entity(HandlerId.Boards, '__TEMP__');
    while (currentEntity.encode() !== firstEntity.encode()) {
      const {
        success,
        entity,
      } = await this.processOne();

      entities.push({
        success,
        entity,
      });

      currentEntity = entity;
    }

    return entities;
  }
}

File: v2/app/src/fourChan/refresher/refresher.go
package refresher

import (
	"fmt"

	redisT "github.com/go-redis/redis/v8"

	"github.com/Allypost/thread-puller/app/providers/redis"
	baseFetcher "github.com/Allypost/thread-puller/app/src/base/fetcher"
	"github.com/Allypost/thread-puller/app/src/fourChan/fetcher"
	"github.com/Allypost/thread-puller/app/src/fourChan/refresher/entity"
	"github.com/Allypost/thread-puller/app/src/fourChan/refresher/handlerId"
	"github.com/Allypost/thread-puller/app/src/fourChan/types/public"
	"github.com/Allypost/thread-puller/app/util/logger"
)

type Handler func(args ...string) (baseFetcher.Encodable, error)

const RedisKey = "refreshers:4chan:queue"

var handlers = make(map[handlerId.HandlerId]Handler)

var log = logger.For("refresher")

func Init() {
	handlers[handlerId.Boards] = func(args ...string) (baseFetcher.Encodable, error) {
		return fetcher.BoardFetcher().RefreshCached()
	}
	handlers[handlerId.Threads] = func(args ...string) (baseFetcher.Encodable, error) {
		return fetcher.ThreadFetcher().RefreshCached(args[0])
	}
}

func Clean() *redisT.Cmd {
	r := redis.RedisProvider()

	log.Debugf("[4chan Refresher] Cleaning queue [%s]", RedisKey)

	return r.Do("del", RedisKey)
}

func Populate() error {
	log.Debugf("[4chan Refresher] Populating queue [%s]", RedisKey)
	boards, err := handlers[handlerId.Boards]()
	if err != nil {
		return fmt.Errorf("failed to get boards: %s", err)
	}

	Push(entity.New(handlerId.Boards, "all"))
	log.Debugf("[4chan Refresher] Added boards overview")

	for _, b := range boards.(public.Boards) {
		Push(entity.New(handlerId.Threads, b.Board))
	}
	log.Debugf("[4chan Refresher] Added all boards overview")

	return nil
}

func ProcessOne() (any, error) {
	ent := Pop()
	if ent == nil {
		return nil, fmt.Errorf("no entity to process")
	}

	res, err := handle(ent)
	if err != nil {
		return nil, err
	}

	Push(*ent)

	return res, nil
}

func Push(entity entity.Entity) *redisT.Cmd {
	r := redis.RedisProvider()

	return r.Do("rpush", RedisKey, entity.String())
}

func Pop() *entity.Entity {
	r := redis.RedisProvider()

	data := r.Do("lpop", RedisKey)

	if data == nil {
		return nil
	}

	resp, err := data.Result()
	if err != nil {
		return nil
	}
	item, err := entity.Decode(resp.(string))
	if err != nil {
		return nil
	}

	return item
}

func handle(entity *entity.Entity) (any, error) {
	handlerId := entity.HandlerId
	entityId := entity.EntityId
	handler := handlers[handlerId]

	if handler == nil {
		return nil, fmt.Errorf("no handler for entity: %s", entity)
	}

	return handler(entityId)
}
